tests/e2e/json_utils.go

```diff
@@ -0,0 +1,375 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+)
+
+// stores a proposal as a raw json, together with its type
+type ProposalAndType struct {
+	RawProposal json.RawMessage
+	Type        string
+}
+
+type (
+	// to have a ChainState object that does not have the overridden Marshal/Unmarshal method
+	ChainStateCopy ChainState
+
+	// duplicated from the ChainState with a minor change to the Proposals field
+	ChainStateWithProposalTypes struct {
+		ChainStateCopy
+		Proposals *map[uint]ProposalAndType // the only thing changed from the real ChainState
+	}
+)
+
+// MarshalJSON marshals a step into JSON while including the type of the action.
+func (step Step) MarshalJSON() ([]byte, error) {
+	actionType := reflect.TypeOf(step.Action)
+
+	result := struct {
+		ActionType string
+		Action     interface{}
+		State      State
+	}{
+		ActionType: actionType.String(),
+		Action:     step.Action,
+		State:      step.State,
+	}
+
+	return json.Marshal(result)
+}
+
+// UnmarshalJSON unmarshals a step from JSON while including the type of the action.
+func (step *Step) UnmarshalJSON(data []byte) error {
+	var tmp struct {
+		ActionType string
+		Action     json.RawMessage
+		State      State
+	}
+	print(string(data))
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+
+	action, err := UnmarshalMapToActionType(tmp.Action, tmp.ActionType)
+	if err != nil {
+		return err
+	}
+
+	step.Action = action
+	step.State = tmp.State
+	return nil
+}
+
+// UnmarshalMapToActionType takes a JSON object and an action type and marshals into an object of the corresponding action.
+func UnmarshalMapToActionType(rawAction json.RawMessage, actionTypeString string) (interface{}, error) {
+	var err error
+	switch actionTypeString {
+	case "main.submitConsumerAdditionProposalAction":
+		var a submitConsumerAdditionProposalAction
+		err = json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.SendTokensAction":
```

Comment:
Similar interface idea could be used here for remembering to include a case for any new action

Reply:
I think this is more or less ok: This will break non-silently when you try to use the action in a trace, so I think it's ok